File: python/dispatch-sdk/src/dispatch_sdk/dedup.py
"""Prevent the same exception object from being reported twice (e.g. by both a framework
handler and a global handler). We tag the exception with a marker attribute, exactly like the
gem's @__dispatch_captured instance variable — builtin exception instances are not always
weak-referenceable (notably on CPython 3.14), so a WeakSet is not reliable here."""

from __future__ import annotations

from typing import Any

_MARKER = "_dispatch_captured"


def already_captured(exc: Any) -> bool:
    return getattr(exc, _MARKER, False) is True


def mark_captured(exc: Any) -> None:
    try:
        setattr(exc, _MARKER, True)
    except (AttributeError, TypeError):
        # Objects with __slots__ and no matching slot can't take the marker; skip dedup
        # for those rather than fail.
        pass

File: packages/core/src/dedup.ts
// Prevents the same error object from being reported twice (e.g. once by a framework error
// handler and again by a global handler after re-raise). The gem sets an instance variable on
// the exception; JS can't add hidden props safely across realms, so we track identity in a
// WeakSet — entries are GC'd with the error, so this never leaks.
const captured = new WeakSet<object>();

function isObject(value: unknown): value is object {
  return typeof value === "object" && value !== null;
}

export function alreadyCaptured(error: unknown): boolean {
  return isObject(error) && captured.has(error);
}

export function markCaptured(error: unknown): void {
  if (isObject(error)) captured.add(error);
}
